docs/advanced_topics.rst

```diff
@@ -1,114 +1,244 @@
+workloads. You can set the DALI CPU thread affinity by using the ``DALI_AFFINITY_MASK`` environment
+variable, which is a comma-separated list of CPU IDs that will be assigned to corresponding DALI
+threads. The number of DALI threads is set during the pipeline construction by the num_threads
+argument and set_affinity enables thread affinity for the CPU worker threads.
+
+.. note::
+  For performance reasons, the hybrid :meth:`nvidia.dali.ops.ImageDecoder` operator, which is
+  nvJPEG based, creates threads on its own, and these threads are always affined.
+
+In ``DALI_AFFINITY_MASK``, if the number of threads is higher than the number of CPU IDs,
+the following process is applied:
+
+1) The threads are assigned to the CPU IDs until all of the CPU IDs from ``DALI_AFFINITY_MASK``
+   are assigned.
+2) For the remaining threads, the CPU IDs from nvmlDeviceGetCpuAffinity will be used.
+
+An example:
 
 .. code-block:: bash
 
   num_threads=5
   DALI_AFFINITY_MASK=3,5,6,10
 
-this will set thread 0 to CPU 3, thread 1 to CPU 5, thread 2 to CPU 6, thread 3 to CPU 10 and thread 4 to CPU id that is returned by nvmlDeviceGetCpuAffinity.
-
+This example sets thread 0 to CPU 3, thread 1 to CPU 5, thread 2 to CPU 6, thread 3 to CPU 10,
+and thread 4 to the CPU ID that is returned by nvmlDeviceGetCpuAffinity.
 
-Memory consumption
+Memory Consumption
 ------------------
 
-DALI uses three kinds of memory: host, host page-locked (pinned) and GPU.
-GPU and host pinned memory allocation and freeing require device synchronization. For this reason, DALI avoids reallocating these kinds of memory whenever possible. The buffers allocated with this kind of storage will only grow when the existing buffer is too small to accommodate the requested shape. This allocation strategy reduces the number of total memory management operations and greatly increases the processing speed after the allocations have stabilized.
-
-In contrast, ordinary host memory is relatively cheap to allocate and free. To reduce host memory consumption, the buffers may shrink if the new requested size is smaller than a specified fraction of the old size (called shrink threshold). It can be adjusted to any value between 0 (never shrink) and 1 (always shrink), with the default being 0.9. The value can be controlled either via environment variable `DALI_HOST_BUFFER_SHRINK_THRESHOLD` or set in Python using `nvidia.dali.backend.SetHostBufferShrinkThreshold` function.
-
-During processing, DALI works on batches of samples. For GPU and some CPU Operators, the batch is stored as contiguous allocation which is processed in one go, which reduces the number of necessary allocations.
-For some CPU Operators, that cannot calculate their output size ahead of time, the batch is instead stored as a vector of separately allocated samples.
+DALI uses the following memory types:
+
+- Host
+- Host-page-locked
+- GPU
+
+Allocating and freeing the GPU and host page-locked (or pinned) memory require
```

Comment:
```suggestion
Allocating and freeing GPU and host page-locked (or pinned) memory require
```
It's unspecified "GPU memory", with GPU being used as an adjective. No article required.

Reply:
Done